scipy/stats/unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,869 @@
+        generator.
+        """
+        return unur_tdr_get_sqhratio(self.rng)
+
+    @property
+    def hat_area(self):
+        """Get the area below the hat for the generator."""
+        return unur_tdr_get_hatarea(self.rng)
+
+    @property
+    def squeeze_area(self):
+        """Get the area below the squeeze for the generator."""
+        return unur_tdr_get_squeezearea(self.rng)
+
+    cdef int _ppf_hat(self, const double *u, double *out, size_t N) except -1:
+        cdef:
+            ccallback_t callback
+            unuran_callback_t unur_callback
+            size_t i
+        init_unuran_callback(&callback, &unur_callback, self.callbacks,
+                             self.params)
+        if setjmp(callback.error_buf) != 0:
+            release_unuran_callback(&callback, &unur_callback)
+            return -1
+        for i in range(N):
+            out[i] = unur_tdr_eval_invcdfhat(self.rng, u[i], NULL, NULL, NULL)
+        release_unuran_callback(&callback, &unur_callback)
+        return 0
+
+    def ppf_hat(self, u):
+        """
+        Evaluate the inverse of the CDF of the hat distribution at u.
+
+        This call does not work for variant IA (immediate acceptance).
+        In this case the hat CDF is evaluated as if variant PS is used.
+
+        Parameters
+        ----------
+        u : array_like
+            An array of percentiles
+
+        Returns
+        -------
+        ppf_hat : array_like
+            Array of quantiles corresponding to the given percentiles.
+
+        Examples
+        --------
+        >>> from scipy.stats import TransformedDensityRejection
+        >>> from scipy.stats import norm
+        >>> from math import pi, sqrt, exp
+        >>> 
+        >>> pdf = lambda x: 1./sqrt(2*pi) * exp(-0.5 * x**2)
+        >>> dpdf = lambda x: -x/sqrt(2*pi) * exp(-0.5 * x**2)
+        >>> rng = TransformedDensityRejection(pdf, dpdf, seed=123)
+        >>> 
+        >>> rng.ppf_hat(0.5)
+        -0.00018050266342393984
+        >>> norm.ppf(0.5)
+        0.0
+        >>> u = np.linspace(0, 1, num=1_000)
+        >>> ppf_hat = rng.ppf_hat(u)
+        """
+        u = np.asarray(u, dtype='d')
+        oshape = u.shape
+        N = np.prod(oshape)
+        cdef double[::1] u_view = u.flatten()
+        cdef double[::1] out = np.empty_like(u_view)
+        if N == 0:
+            return np.asarray(out).reshape(oshape)
+        self._ppf_hat(&u_view[0], &out[0], N)
+        if u.ndim == 0:
+            return out[0]
+        return np.asarray(out).reshape(oshape)
+
+
+cdef class DiscreteAliasUrn(Method):
+    r"""
+    Discrete Alias-Urn Method.
+
+    This method is used to sample from discrete distributions with a finite
+    domain. It uses the probability vector of size :math:`N` or a probability
+    mass function with a finite support to generator random numbers from the
+    distribution.
+
+    Parameters
+    ----------
+    pv_or_pmf : array_like, callable
```

Comment:
array_like or callable

Reply:
This has been changed while adding the `dist` parameter.